radar-commons-android/src/main/java/org/radarcns/android/auth/portal/MetaTokenParser.java

```diff
@@ -0,0 +1,44 @@
+package org.radarcns.android.auth.portal;
+
+import android.support.annotation.NonNull;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.radarcns.android.auth.AppAuthState;
+import org.radarcns.android.auth.AuthStringParser;
+
+import java.io.IOException;
+
+import static org.radarcns.android.auth.portal.ManagementPortalClient.BASE_URL_PROPERTY;
+import static org.radarcns.android.auth.portal.ManagementPortalClient.MP_REFRESH_TOKEN_PROPERTY;
+import static org.radarcns.android.auth.portal.ManagementPortalClient.PRIVACY_POLICY_URL_PROPERTY;
+
+/**
+ * Reads refreshToken and meta-data of token from json string and sets it as property in
+ * {@link AppAuthState}.
+ */
+public class MetaTokenParser implements AuthStringParser {
+
+    private AppAuthState currentState;
+
+    public MetaTokenParser(AppAuthState state) {
+        this.currentState = state;
+    }
+
+    @Override
+    public AppAuthState parse(@NonNull String authString) throws IOException {
+        String refreshToken = (String) currentState.getProperty(MP_REFRESH_TOKEN_PROPERTY);
+
+        try {
+            JSONObject json = new JSONObject(authString);
+            refreshToken = json.optString("refreshToken", refreshToken);
```

Comment:
if the string is certainly there, you can use `getString` here and remove the earlier line to get it from the current state.